standard/expressions.md

```diff
@@ -37,6 +37,155 @@ Most of the constructs that involve an expression ultimately require the express
 
+### §patterns-new-clause Patterns and pattern matching
+
+#### §patterns-new-clause-general General
+
+A ***pattern*** is a syntactic form that can be used with the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)) and in a *switch_statement* ([§12.8.3](statements.md#1283-the-switch-statement)) to express the shape of data against which incoming data is to be compared. A pattern is tested in the context of a switch expression or a *relational_expression* that is on the left-hand side of an `is` operator. Let us call this a ***pattern input value***.
+
+A pattern may have one of the following forms:
+
+```ANTLR
+pattern
+    : declaration_pattern
+    | constant_pattern
+    | var_pattern
+    ;
+```
+
+A *declaration_pattern* and a *var_pattern* can result in the declaration of a local variable. The scope of such a variable is as follows:
+
+- If the pattern is a case label ([§12.8.3](statements.md#1283-the-switch-statement)), then the scope of the variable is the associated *case block*.
+- Otherwise, the variable is part of the pattern that is the right-hand operand of the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)), and the variable’s scope is based on the construct immediately enclosing the `is` expression containing the pattern, as follows:
+  - If the expression is in an expression-bodied lambda, the variable's scope is the body of the lambda.
+  - If the expression is in an expression-bodied method or property, the variable's scope is the body of the method or property.
+  - If the expression is in a `when` clause of a `catch` clause, the variable's scope is that `catch` clause.
+  - If the expression is in an *iteration_statement*, the variable's scope is just that statement.
+  - If the expression is in a *constructor_initializer*, the variable's scope is the part of that *constructor_initializer* following the expression, and the body of the constructor.
+  - If the expression is in a *variable_initializer* of a field, the variable's scope is the part of that *variable_initializer* following the expression.
+  - If the expression is in a *query_expression* that is translated into the body of a lambda, the variable's scope is that *query_expression*.
+  - Otherwise if the expression is in some other statement form, the variable's scope is the scope containing the statement.
+
+For the purpose of determining the scope, an *embedded_statement* is considered to be in its own scope.
+
+#### §declaration-pattern-new-clause Declaration pattern
+
+A *declaration_pattern* is used to test that a value has a given type and if the test succeeds, to cast that value to that type.
```

Comment:
We should introduce the terminology of a pattern possibly *matching* a value, and use it consistently.

Reply:
Maybe *matching* is trying, and it *succeeds* or *fails*?